src/libraries/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyDependencyResolver.cs

```diff
@@ -95,7 +105,16 @@ public AssemblyDependencyResolver(string componentAssemblyPath)
             _assemblyPaths = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
             foreach (string assemblyPath in assemblyPaths)
             {
-                _assemblyPaths.Add(Path.GetFileNameWithoutExtension(assemblyPath), assemblyPath);
+                // Check to see if we have the same managed assembly simple name in the dictionary.
+                // We compare the paths reported and if they're the same, simply ignore the duplicate entries. And only fail if the paths are different
+                // (since hostpolicy normalizes paths, the comparison should be sufficient to perform as simple string case sensitive/insensitive comparison)
+                // On Windows - OrdinalIgnoreCase comparison
+                // Elsewhere - Ordinal comparison
+                string assemblySimpleName = Path.GetFileNameWithoutExtension(assemblyPath);
+                if (!_assemblyPaths.ContainsKey(assemblySimpleName) || !_assemblyPaths[assemblySimpleName].Equals(assemblyPath, StringComparisonMode))
```

Comment:
`ContainsKey` here is redundant lookup for the common case. If we decide that we want to have some error handling for collisions, it would be better to call `TryAdd` and only do the additional error handling work if there is a collision:

```
if (!_assemblyPaths.TryAdd(Path.GetFileNameWithoutExtension(assemblyPath), assemblyPath))
{
    decide whether to throw exception or not
}
```

Reply:
I'll chat with @vitek-karas more to understand the existing tests and to see if we can add a case to cover this scenario

Reply:
Added tests and also opened issue #42334 to track some behavioral differences between mac-Linux host policy